File: 278_FirstBadVersion.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        left, right = 1, n
        while left < right:
            mid = (left + right)>>1
            
            if not isBadVersion(mid):
                left = mid+1
            else:
                if not isBadVersion(mid-1):
                    return mid
                else:
                    right = mid-1
        return right

File: cpp/278_FirstBadVersion.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int left = 0, right = n;
        while (left < right) {
            int mid = left + (right - left) / 2;
            bool bad = isBadVersion(mid);
            
            if (bad) right = mid;
            else left = mid + 1;
        }
        return left;
    }
};
